src/App.js

```diff
@@ -23,49 +23,108 @@ const generateSquares = () => {
+  const [currentPlayer, setCurrentPlayer] = useState(PLAYER_1);
 
   // Wave 2
-  // You will need to create a method to change the square 
+  // You will need to create a method to change the square
   //   When it is clicked on.
   //   Then pass it into the squares as a callback
-
+  const onClickCallback = (clickedSquare) => {
+    const newSquares = [...squares];
+    for (let i = 0; i < 3; i++) {
+      for (let j = 0; j < 3; j++) {
+        if (checkForWinner() === null) {
+          if (newSquares[i][j].id === clickedSquare.id) {
+            if (newSquares[i][j].value === '') {
+              newSquares[i][j].value = currentPlayer;
+              if (currentPlayer === PLAYER_1) {
+                setCurrentPlayer(PLAYER_2);
+              } else {
+                setCurrentPlayer(PLAYER_1);
+              }
+            }
+          }
+        }
+      }
+    }
+    setSquares(newSquares);
+    checkForWinner();
+  };
 
   const checkForWinner = () => {
-    // Complete in Wave 3
-    // You will need to:
-    // 1. Go accross each row to see if 
-    //    3 squares in the same row match
-    //    i.e. same value
-    // 2. Go down each column to see if
-    //    3 squares in each column match
-    // 3. Go across each diagonal to see if 
-    //    all three squares have the same value.
+    let i = 0;
 
-  }
+    // Check all the rows and columns for a winner
+    while (i < 3) {
+      if (
+        squares[i][0].value === squares[i][1].value &&
+        squares[i][2].value === squares[i][1].value &&
+        squares[i][0].value !== ''
+      ) {
+        return squares[i][0].value;
+      } else if (
+        squares[0][i].value === squares[1][i].value &&
+        squares[2][i].value === squares[1][i].value &&
+        squares[0][i].value !== ''
+      ) {
+        return squares[0][i].value;
+      }
+      i += 1;
+    }
+    // Check Top-Left to bottom-right diagonal
+    if (
+      squares[0][0].value === squares[1][1].value &&
+      squares[2][2].value === squares[1][1].value &&
+      squares[1][1].value !== ''
+    ) {
+      return squares[0][0].value;
+    }
+
+    // Check Top-right to bottom-left diagonal
+    if (
+      squares[0][2].value === squares[1][1].value &&
+      squares[2][0].value === squares[1][1].value &&
+      squares[1][1].value !== ''
+    ) {
+      return squares[0][2].value;
+    }
+
+    return null;
+  };
 
   const resetGame = () => {
     // Complete in Wave 4
-  }
+    setSquares(generateSquares());
+    setCurrentPlayer(PLAYER_1);
+  };
+
+  const changeAppHeader = () => {
+    if (checkForWinner() === null) {
+      return `Current Player ${currentPlayer}`;
+    } else {
+      return `Winner is ${checkForWinner()}`;
+    }
+  };
 
   return (
     <div className="App">
       <header className="App-header">
         <h1>React Tic Tac Toe</h1>
-        <h2>The winner is ... -- Fill in for wave 3 </h2>
-        <button>Reset Game</button>
+        <h2>{changeAppHeader()}</h2>
```

Comment:
Nice helper function! Another approach to displaying winner text is by using a ternary operator:

```suggestion
        <h2> { winner ? Winner is ${winner} : `Current Player ${currentPlayer}` } </h2>
```